image_proc/src/crop_non_zero.cpp

```diff
@@ -0,0 +1,149 @@
+*     from this software without specific prior written permission.
+* 
+*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+*  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+*  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+*  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+*  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+*  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+*  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+*  POSSIBILITY OF SUCH DAMAGE.
+*********************************************************************/
+#include <rclcpp/rclcpp.hpp>
+#include <image_transport/image_transport.h>
+#include <boost/thread.hpp>
+#include <cv_bridge/cv_bridge.h>
+#include "crop_non_zero.hpp"
+//#include <algorithm> // for std::max_element
+
+namespace image_proc
+{
+
+namespace enc = sensor_msgs::image_encodings;
+
+CropNonZeroNode::CropNonZeroNode(const rclcpp::NodeOptions& options) : Node("CropNonZeroNode", options)
+{
+  auto parameter_change_cb =
+    [this](std::vector<rclcpp::Parameter> parameters) -> rcl_interfaces::msg::SetParametersResult {
+    auto result = rcl_interfaces::msg::SetParametersResult();
+    result.successful = true;
+    for (auto parameter : parameters)
+    {
+      if (parameter.get_name() == "camera_namespace")
+      {
+        camera_namespace_ = parameter.as_string();
+        RCLCPP_INFO(get_logger(), "camera_namespace: %s ", camera_namespace_.c_str());
+        break;
+      }
+    }
+    image_pub_ = camera_namespace_ + "/image";
+    connectCb();
+    // Make sure we don't enter connectCb() between advertising and assigning to pub_depth_
+    std::lock_guard<std::mutex> lock(connect_mutex_);
+    RCLCPP_INFO(this->get_logger(), "publish: %s", image_pub_.c_str());
+    pub_ = image_transport::create_publisher(this, image_pub_);
+    return result;
+  };
+  this->declare_parameter("camera_namespace");
+
+  rclcpp::Parameter parameter;
+  if (rclcpp::PARAMETER_NOT_SET != this->get_parameter("camera_namespace", parameter)) {
+    parameter_change_cb(this->get_parameters({"camera_namespace"}));
+  }
+
+  this->set_on_parameters_set_callback(parameter_change_cb);
+}
+
+// Handles (un)subscribing when clients (un)subscribe
+void CropNonZeroNode::connectCb()
+{
+  image_sub_ = camera_namespace_ + "/image_raw";
+  std::lock_guard<std::mutex> lock(connect_mutex_);
+
+  RCLCPP_INFO(this->get_logger(), "subscribe: %s", image_sub_.c_str());
+  sub_raw_ = image_transport::create_subscription(this, image_sub_, std::bind(&CropNonZeroNode::imageCb, this, std::placeholders::_1), "raw");
```

Comment:
Should the type of compression be a parameter? (actually asking, I'm not sure what the ROS1 version did)

Reply:
Seems it's just hard-coded: 

```    
image_transport::TransportHints hints("raw", ros::TransportHints(), getPrivateNodeHandle());

sub_raw_ = it_->subscribe("image_raw", 1, &CropNonZeroNodelet::imageCb, this, hints);
```

Reply:
Eh. Not ideal but not going to make it part of this PR